packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1918 @@
+   * @param specification - The specification of the permission.
+   * @param permission - The permission to validate.
+   * @param origin - The origin associated with the permission.
+   * @param targetName - The target name of the permission.
+   */
+  private validatePermission(
+    specification: PermissionSpecificationConstraint,
+    permission: PermissionConstraint,
+    origin: OriginString,
+    targetName: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+  ): void {
+    const { allowedCaveats, validator } = specification;
+    if (permission.caveats !== null) {
+      permission.caveats.forEach(({ type }) => {
+        if (!allowedCaveats?.includes(type)) {
+          throw new ForbiddenCaveatError(type, origin, targetName);
+        }
+      });
+    }
+
+    if (validator) {
+      validator(permission, origin, targetName);
+    }
+  }
+
+  /**
+   * Assigns the specified permissions to the subject with the given origin.
+   * Overwrites all existing permissions, and creates a subject entry if it
+   * doesn't already exist.
+   *
+   * **ATTN:** Assumes that the new permissions have been validated.
+   *
+   * @param origin - The origin of the grantee subject.
+   * @param permissions - The new permissions for the grantee subject.
+   */
+  private setValidatedPermissions(
+    origin: OriginString,
+    permissions: Record<
+      string,
+      ExtractPermission<
+        ControllerPermissionSpecification,
+        ControllerCaveatSpecification
+      >
+    >,
+  ): void {
+    this.update((draftState) => {
+      if (!draftState.subjects[origin]) {
+        draftState.subjects[origin] = { origin, permissions: {} };
+      }
+
+      // Typecast: Immer WritableDraft incompatibility
+      draftState.subjects[origin].permissions = castDraft(permissions);
+    });
+  }
+
+  /**
+   * Constructs the requested caveats for the permission of the specified
+   * subject origin and target name.
+   *
+   * Throws an error if validation fails.
+   *
+   * @see {@link PermissionController.constructCaveat} for details.
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveats - The requested caveats to construct.
+   * @returns The constructed caveats.
+   */
+  private constructCaveats(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveats?: unknown[] | null,
+  ): NonEmptyArray<ExtractCaveats<ControllerCaveatSpecification>> | undefined {
+    const caveatArray = requestedCaveats?.map((requestedCaveat) =>
+      this.constructCaveat(origin, target, requestedCaveat),
+    );
+    return caveatArray && caveatArray.length > 0
+      ? (caveatArray as NonEmptyArray<
+          ExtractCaveats<ControllerCaveatSpecification>
+        >)
+      : undefined;
```

Comment:
We should probably create a utility function that uses a type predicate to assert an array is non-empty, rather than casting each time like this.

e.g. this should do the trick:

```
function isNonEmptyArray<T>(thing: T[]): thing is NonEmptyArray<T> {
  return Array.isArray(thing) && thing.length > 0;
}
```

Then in this case, we don't need the cast anymore:

```suggestion
    return caveatArray && isNonEmptyArray(caveatArray)
      ? caveatArray
      : undefined;
```

Reply:
Good suggestion! 5ce5f80